src/main/java/com/google/devtools/build/lib/rules/repository/WorkspaceFileHelper.java

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.rules.repository;
+
+import com.google.devtools.build.lib.actions.FileValue;
+import com.google.devtools.build.lib.cmdline.LabelConstants;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import com.google.devtools.build.lib.vfs.Root;
+import com.google.devtools.build.lib.vfs.RootedPath;
+import com.google.devtools.build.skyframe.SkyFunction.Environment;
+import com.google.devtools.build.skyframe.SkyKey;
+
+import java.io.IOException;
+
+public class WorkspaceFileHelper {
+
+  public static RootedPath getWorkspaceRootedFile(Root directory, Environment env)
+      throws IOException, InterruptedException {
+    return getWorkspaceRootedFile(RootedPath.toRootedPath(directory, PathFragment.EMPTY_FRAGMENT), env);
+  }
+
+  /**
+   * Get a RootedPath of the WORKSPACE file we should use for a given directory.
+   * @param directory The directory that could contain WORKSPACE.bazel or WORKSPACE file.
+   * @param env Skyframe env
+   * @return A RootedPath to <directory>/WORKSPACE.bazel file if it exists and it's not a directory,
+   *         otherwise, return a RootedPath to <directory>/WORKSPACE file.
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public static RootedPath getWorkspaceRootedFile(RootedPath directory, Environment env)
+      throws IOException, InterruptedException {
+    RootedPath workspaceRootedFile =
+        RootedPath.toRootedPath(
+            directory.getRoot(),
+            directory
+                .getRootRelativePath()
+                .getRelative(LabelConstants.WORKSPACE_DOT_BAZEL_FILE_NAME));
+
+    SkyKey workspaceFileKey = FileValue.key(workspaceRootedFile);
+    FileValue value;
+    value = (FileValue) env.getValueOrThrow(workspaceFileKey, IOException.class);
+    if (value == null) {
+      return null;
+    }
+
+    if (value.exists() && !value.isDirectory()) {
+      return workspaceRootedFile;
```

Comment:
Returning here means `WORKSPACE.bazel` exists and it's not a directory.

Returning in line 63 means `WORKSPACE.bazel` is missing (or it's a directory), but says nothing about `WORKSPACE` because it's not checked.

So the caller can't assume the returned path exists and must request it again from Skyframe. (Which Skyframe might even report as an error.)

So the caller won't know if the file exists or not.

Maybe request the second FileValue as well, and return the RootedPath + FileValue pair?

Reply:
Indeed, the caller shouldn't assume the returned path exists. This function just returns the path of the workspace file, then the caller should request the value again. In some case, it would be FileValue, but other cases, it would be `PackageValue` or `WorkspaceFileValue`. So requesting the FileValue of `WORKSPACE` is actually unnecessary.

I'll add a comment to make this clear, WDYT?

Reply:
If this function were inlined, then such callers would just request the PackageValue or WorkspaceFileValue for "WORKSPACE.bazel" and if missing then for "WORKSPACE", without ever requesting a FileValue, right? Because presumable PackageValue and WorkspaceFileValue already depend on the FileValue.

So this logic is actually not generalizable, and documenting and understanding the fine details is more effort than duplicating the code and slightly adjusting it as needed at its call sites. So I recommend inlining this method after all.

Reply:
There are actually three steps in the logic at every places this function gets called.

1. Get FileValue of `WORKSPACE.bazel`

2. Check the FileValue to determine we should use `WORKSPACE` or `WORKSPACE.bazel`

3. Get FileValue/PackageValue/WorkspaceFileValue of the given workspace file.

As you can see, the first two steps are the same, so they could be refactored out. Inline this function would cause this logic to be implemented at 4 different places. So I guess we should keep this?

Reply:
Thinking more about this, I have no strong argument or better idea. Let's keep it like this.